### PR TITLE
fixes #476: pixel / projected meter / lat/long conversions

### LINKED ISSUE #476: conversion routines
* [x] pixels `<->` projected meters `<->` coordinates
 * [x] meters/pixel at a given zoom & latitude

Precursors to point plotting #60 and interactivity #352. 

Related: #97 

Happening in [`conversions`](https://github.com/mapbox/mapbox-gl-native/tree/conversions).

### PR DESCRIPTION
**Pull request by incanus**
Remaining: 

* [x] Pick up `conversions` work again and merge in `master` (over 1,000 commits).

* [x] Integrate old `gl-cocoa#conversions` changes into this repo.

* [x] Reduce overlap between `Transform` and `TransformState` as a result of this branch + https://github.com/mapbox/mapbox-gl-native/commit/4603ca037cab2d0f732285ce9e006744ad3118d6 (background image support /cc @jfirebaugh).

* [ ] Write specific tests for exercising conversion routines (will hit iOS specifically in #830). 

* [x] Merge this in (my excuse to finally try rebasing). 

* [x] Delete old [`gl-cocoa#conversions`](https://github.com/mapbox/mapbox-gl-cocoa/tree/conversions) to clean house.

### CODE REVIEW

**Comment by incanus** [comment]
Ok, I think we're set here, including migrating API across OS X, iOS, Linux, and Android. 

Working on tests now.

**Comment by ljbade** [comment]
:-D

File: include/mbgl/map/map.hpp [hunk 1]
```diff
@@ -106,8 +108,8 @@ class Map : private util::noncopyable {
     double getScale() const;
     void setZoom(double zoom, std::chrono::steady_clock::duration duration = std::chrono::steady_clock::duration::zero());
     double getZoom() const;
-    void setLonLatZoom(double lon, double lat, double zoom, std::chrono::steady_clock::duration duration = std::chrono::steady_clock::duration::zero());
-    void getLonLatZoom(double &lon, double &lat, double &zoom) const;
+    void setLatLngZoom(LatLng latLng, double zoom, std::chrono::steady_clock::duration duration = std::chrono::steady_clock::duration::zero());
+    void getLatLngZoom(LatLng &latLng, double &zoom) const;
```
Comment (jfirebaugh) [inline include/mbgl/map/map.hpp]: Can we remove this method and use separate calls to `getLatLng()` and `getZoom()`?

File: include/mbgl/map/map.hpp [hunk 2]
```diff
@@ -127,6 +129,15 @@ class Map : private util::noncopyable {
     void setAccessToken(const std::string &token);
     const std::string &getAccessToken() const;
 
+    // Projection
+    inline void getWorldBoundsMeters(ProjectedMeters &sw, ProjectedMeters &ne) const { Projection::getWorldBoundsMeters(sw, ne); }
```
Comment (jfirebaugh) [inline include/mbgl/map/map.hpp]: Instead of passing out args via reference, there should be a `template <class T> class Bounds` and the return type of this function should be `Bounds<ProjectedMeters>`.

File: include/mbgl/map/map.hpp (hunk 1 above)
Comment (incanus) [inline include/mbgl/map/map.hpp]: Yeah, I think so. This was just ported for historical reasons. Good reason to scrap it.

File: include/mbgl/map/transform.hpp [hunk 3]
```diff
@@ -52,6 +53,10 @@ class Transform : private util::noncopyable {
     void startRotating();
     void stopRotating();
 
+    // Projection
+    void pixelForLatLng(const LatLng latLng, double &x, double &y) const;
```
Comment (jfirebaugh) [inline include/mbgl/map/transform.hpp]: Should return `vec2<double>`.

File: src/mbgl/renderer/painter.hpp [hunk 4]
```diff
@@ -178,7 +177,7 @@ class Painter : private util::noncopyable {
     }();
 
 private:
-    TransformState state;
+    Transform& transform;
```
Comment (jfirebaugh) [inline src/mbgl/renderer/painter.hpp]: Can we revert the changes to painter.hpp/cpp and painter_*.cpp? We're trying to reduce shared state through references, and using `Transform&` reverses that.

Comment (incanus) [inline src/mbgl/renderer/painter.hpp]: Yeah, that's my one worry about this PR, the way we "poke through" the transform to grab the state. The painter needs to know the conversion (as put in originally in your background stuff in 4603ca0), but perhaps this can be pushed down to the state, then have the public map APIs poke through map to transform to transform state to get them. Will look at this.

Comment (jfirebaugh) [inline src/mbgl/renderer/painter.hpp]: If it helps, I think it's fine to have `Painter::render` take a `Transform` by copy rather than a `TransformState`, so that this member becomes `Transform transform`. In that case I would still keep a `TransformState& state` member around that references `transform.currentState()` to avoid needing to change all the paint implementation methods.

Long term I think `Transform` and `TransformState` will merge.

Comment (incanus) [inline src/mbgl/renderer/painter.hpp]: `Transform` is non-copyable?

File: include/mbgl/map/map.hpp (hunk 1 above)
Comment (incanus) [inline include/mbgl/map/map.hpp]: 429f7cf

File: include/mbgl/map/transform.hpp (hunk 3 above)
Comment (incanus) [inline include/mbgl/map/transform.hpp]: Reason I did this was because calling code (e.g. iOS) would have to know about `vec2` type in this case.

Comment (incanus) [inline include/mbgl/map/transform.hpp]: Also isn't there an implicit `x, y` ordering that would be lost?

Comment (incanus) [inline include/mbgl/map/transform.hpp]: f13ea06803667d12bf501b8a98865603001bde95

File: src/mbgl/renderer/painter.hpp (hunk 4 above)
Comment (incanus) [inline src/mbgl/renderer/painter.hpp]: I'm remembering why I did this now — this would require moving `transform.getLatLng()` down to the state as well, along with the spherical mercator math cache values. Seemed intrusive, but perhaps less so than the painter stuff. Trying this now.

Comment (incanus) [inline src/mbgl/renderer/painter.hpp]: c2a9adb13319b18e54f4c2637a7bbdd0f8283198

**Comment by incanus** [comment]
This is now working, so I'm inclined to merge it once it passes Travis and write tests in `master`.
